### DISCUSSION TITLE
First attempt with Ctrlr (Waldorf Blofeld panel)

### DISCUSSION THREAD

**Discussion by Leigten** [RomanKubiak/ctrlr#625] [Q&A] [answered]
Hello everyone
I am trying to make a panel for the Waldorf Blofeld. 
I've never done anything like this and it's my first time using Ctrlr and I'm trying to make something that is as simple as possible. 
So far I have defined the whole oscillators section and everything works great except for one thing. 
The Blofeld has a function to define whether to use it in polyphonic or monophonic mode, called Allocation. Its default value is "poly" (0) or "mono" (1).
Another function, called Unisono, allows you to assign how many voices to unison : off , dual, 3, 4, 5 or 6 voices.
For each of the two functions I have defined an uiCombo button.
The Allocation's one has poly and mono in the Listbox contents, MIDI message type "Sysex" and in the SysEx Formula field the following string : f0 3e 13 00 20 00 00 3a xx f7.
For the Unisono function I put off,dual,3,4,5 and 6 in the Listbox contents, MIDI message type "Sysex" and in the SysEx Formula field the following string : f0 3e 13 00 20 00 00 3a xx f7.
In addition, since the hexadecimal values of xx for Unisono, unlike Allocation, do not start from 00 but from 05 with steps of 16 (thus 05,15,25,35,45,55) I inserted the following epression "modulatorValue*16+5" in the "Expression to evaluate when calculating the midi message value from the modulator value" section.
Looking at the MIDI Monitor, the values transmitted by the two controls created in Ctrlr are the same as those transmitted by the Blofeld by moving the respective controls, 
but since the parameter they are associated with is the same (3a in the SysEx string) when I set Allocation to poly and then Unison to dual or more, the setting of Unison alters that of Allocation and vice versa. Allocation, if it was at poly is then set to mono, 
and if Unison is at any value between dual and 6, it is reset at the time Allocation is set to poly.
In the Waldorf Blofeld manual I read this:

F0h, 3Eh, 13h, DEV, 20h, LL, HH, PP, XX, F7h
where:

0 EXC F0h Marks Start of SysEx
1 IDW 3Eh Waldorf Music ID
2 IDE 13h Blofeld ID
3 DEV Device ID
4 IDM 20h here SNDP (Sound Parameter change)
5 LL see above Location
6 HH see above Parameter index high byte
7 PP see above Parameter index low byte
8 XX see 3.1 New Parameter value
9 EOX F7h End of Exclusive

and then :

4.10 Allocation and Unisono Modes

0uuu000a
uuu:
Index Description

0 off
1 dual
2 3
3 4
4 5
5 6

a:
Index Description

0 Poly
1 Mono

I know that the problem is right here : 0uuu000a 
but I don't know how to overcome it.

![test](https://github.com/RomanKubiak/ctrlr/assets/164783721/d15a91c9-09d0-4c76-8a5e-4b575ddc70e2)

**Comment by Ted-juh** [comment]
Is it a typo or are do you put the Allocation and the unisono at the same spot in the sysex address? 

Edit:
If I can find the sysex manual online, I'll take a look. Because it seems unlikely that unisono and allocation use the same address. Because you state position 5 is Allocation, which is set to 00 in your example. I also wonder if modulatorValue*16+5 is correct but I could be wrong.

**Reply by Leigten** [reply]
It's not a typo, the value is the same for both, unlike all the other parameters. If I move their respective knobs in the Blofeld, I get always the same hexadecimal parameters 3e (as written in the Sysex documentation). This is the very source of the problem. 
Edit: 
I never stated that position 5 in the Sysex string refers to Allocation. Anyway, in the second half of my post I pasted from the Blofeld's sysex manual. The generic Sysex format for sound parameters is the following:
F0h, 3Eh, 13h, DEV, 20h, LL, HH, PP, XX, F7h
What defines the parameter to change and its value are PP (the parameter) and XX (the value). 
As for the modulatorValuex16+5, the hexadecimal values I have to send to the Blofeld are 05,15,25,35,45,55.
The Ctrlr knob sends values from 0 to 5, so I have to transform them with the expression 
modulatorValuex16+5
So:
00x16=0 0+5=5 -> 05 hex
01x16=16 16+5 =21 -> 15 hex
02x16=32 32+5=37 -> 25 hex
03x16=48 48+5=53 -> 35 hex

and so on......

**Reply by Ted-juh** [reply]
You are right, it shares the same address as: 
0uuu000a.

So, unisono mode off with mono would be 00000001. 
Unisono dual mode with mono would be 00100001. 
Unisono 6 with poly would become: 0101000 if I'm correct.

If that's right, then there are 12 possible combinations. Which should be easy to create a Lua method for to read and write the right byte.

**Reply by Leigten** [reply]
Thanks for your replies! 😀
Keep in mind that I started a week ago with Ctrlr and sysex. With much effort I got this far. Everything works except this damn parameter shared by two different functions. Now, for me to be able to define a LUA method is an insurmountable obstacle, since I don't know what LUA is. 😂

**Comment by Ted-juh** [comment]
No worries, we've all been there. We're going to take this slow.

Sorry, I misunderstood "location" for "allocation", I read too fast.

Just one question though. Where did you get the idea that the sysex for the unisono/ poly should start from 5? Can't find that in the sysex manual.

Because to me it's like this:

Unisono off, poly = 0000 0000 Hex 00 Dec 00
Unisono off, mono = 0000 0001 Hex 01 Dec 01

Unisono Dual, poly = 0001 0000 Hex 10 Dec 16
Unisono Dual, mono = 0001 0001 Hex 11 Dec 17

Unisono 3, poly = 0010 0000 Hex 20 Dec 32
Unisono 3, mono = 0010 0001 Hex 21 Dec 33

Unisono 4, poly = 0011 0000 Hex 30 Dec 48
Unisono 4, mono = 0011 001 Hex 31 Dec 49

Unisono 5, poly = 0100 0000 Hex 40 Dec 64
Unisono 5, mono = 0100 0001 Hex 41 Dec 65

Unisono 6, poly = 0101 0000 Hex 50 Dec 80
Unisono 6, mono = 0101 0001 Hex 51 Dec 81

It does explain why you turn poly or mono at certain values or switch unisono with not the desired effect because your modulatorValue*16+5 doesn't output the right values.

Can you send the Hex values from my method in the appropriate sysex format to the Blofeld to see if the result is ok?

**Reply by Leigten** [reply]
I've simply read the sysex strings in the MIDI monitor.
As I told you I started without knowing anything regarding Ctrlr and Sysex, I didn't even own a SysEx manual for the Blofeld.
The manual I've got when I purchased it doesn't go over CC Midi implementation.
I've started inserting knobs in Ctrlr, then read the values received from the Blofeld in the MIDI Monitor and associate the sysex strings to the knobs, replacing the actual values with the variable xx.
All worked, except Allocation and Unisono. Then, I searched for the SysEx manual. I think that the values I've read for the Unisono parameter were wrong, of course. I've sent each of your values to the Blofeld, each one in a single string, and they work correctly.
So, in this way Allocation and Unisono are set simultaneously with a single message. How do we deal with the fact that they are two distinct parameters in the Blofeld, hence two distinct controls in the panel I'm making?

**Reply by Ted-juh** [reply]
Short answer: Lua. Long answer: Too complicated to write a script on my phone.

Can you post the panel here as a zip file?

**Reply by Leigten** [reply]
Here it is, exported as compressed binary, then zipped.
[Waldorf Blofeld_1_0_Test.zip](https://github.com/RomanKubiak/ctrlr/files/14736913/Waldorf.Blofeld_1_0_Test.zip)

**Reply by Ted-juh** [reply]
And just the .panel file? Just in case..

Going to have a look later.

**Reply by Leigten** [reply]
Again, thanks for your replies and your patience!
[Waldorf Blofeld_1_0_Panel.zip](https://github.com/RomanKubiak/ctrlr/files/14736935/Waldorf.Blofeld_1_0_Panel.zip)

**Comment by dnaldoog** [comment, marked as answer]
You say the sysex should be:

00x16=0 0+5=5 -> 05 hex
01x16=16 16+5 =21 -> 15 hex
02x16=32 32+5=37 -> 25 hex
03x16=48 48+5=53 -> 35 hex

&#x2026; but  I would have thought it should be as  as [Tedjuh suggested](https://github.com/RomanKubiak/ctrlr/discussions/625#discussioncomment-8895090). 

If you refer to his binary in his post, you will see that the bits for **Unisono**  __u__  **0uuu000a**, live in bits 4,5,6 and bits for **Allocation** are at bit 0 __a__ and act as a switch. Bits 1,2,3 are disabled and bit 7 must always be 0 in sysex data. combine them and you are limited to values:

00,01,10,11,20,21,30,31,40,41,50,51 (hex)

Can you please check this panel?
[Waldorf Blofeld_1_0_Test_dnaldoog2.zip](https://github.com/RomanKubiak/ctrlr/files/14737384/Waldorf.Blofeld_1_0_Test_dnaldoog2.zip)

I use a global variable `k0` attached to the **Allocation** _listBox_ which adds itself to the sysex sent by **Unisono** and a global variable `k1` attached to the **Unisono** _listBox_ which adds itself to the sysex sent by **Allocation** listBox. 

P.S. If you are squishing down the `listBox` it's probably better to use `uiCombo` :)

**Reply by Ted-juh** [reply]
When I tend to overcomplicate things, there's always a certain Mr Goodland to the rescue. Looking at the panel he posted, the solution is simpler than writing a Lua Script. 

At first, I didn't quite understand where k0 and k1 came from but it must be "Modulator attached to global variable".  That's defined by "k" and then the selected number.  

Neat :)

**Reply by Leigten** [reply]
First of all, thank you for your response and I apologize if I was late in responding. 
I had a health problem and could in no way respond earlier.
Yes, the correct values are definitely those given by Tedjuh.
The ones I said I simply read them into the MIDI monitor input, but for some reason they were not the correct ones (although with those values on the Blofeld the Unison parameters were changing in the correct way....)
Your panel works great! There is only one problem:
I would not be able to deal with such a problem if it came up again (and I do believe there are other parameters that would require such a solution).
As you said : "I use a global variable k0 attached to the Allocation listBox which adds itself to the sysex sent by Unison and a global variable k1 attached to the Unison listBox which adds itself to the sysex sent by Allocation listBox."
But I still don't know how and where to define a variable and how to associate it with the parameters I'm interested in. 
Unfortunately it is my first experience with Ctrlr and System Exclusive, what I managed to do is what you saw by opening my panel. 
Ctrlr's complete lack of documentation doesn't help (especially for many things related to the "Modulator" and "MIDI" sections).
As for your point about using uiCombo rather than uiListBox, I agree with you (in fact for oscillators I used uiCombo).
However, I also like the idea of stretching the listbox to show all the available options, as you did regarding Unisono listbox.
I would not do the same thing for oscillators, but for Unisono there are only 6 options and there is enough space to show them all.

**Reply by Ted-juh** [reply]
It's not like we're busy with Ctrlr 24/7 so no need to apologize. 

Anyways, going over the parameter values in the sysex manual, it looks like only the next one uses a different format:

4.12 Envelope Modes and Triggers
*****************************************************************************
0ttmmmmm
tt:
Index Description
-----------------------------------------------------------------------------
0 normal
1 single
-----------------------------------------------------------------------------
mmmmm:
Index Description
-----------------------------------------------------------------------------
0 ADSR
1 ADS1DS2R
2 One Shot
3 Loop S1S2
4 Loop All
-----------------------------------------------------------------------------
*****************************************************************************
Bit 7 is always 0
It isn't clear from the instructions but "tt" might be 2 bits.  That would make bit 5 and 6 00 and 01. 
Bit 0 - 4  is, conveniently 0-4 

So to send Single/ Loop All: 0 01 00100

Possible Combinations in Hex: 00,01,02,03,04, 40, 41, 42, 43, 44  (To test the values itself)

Now, like I said before, that modulatorValueExpression still confuses the hell out of me, so I have to put up my pouty kitten face to make John weak in the knees and show us how you would set that up in the modulatorValueExpression field.

Apart from that, only the arpeggiator (3.1 SDATA Index number 326 - 347) uses some different formats, but nothing else when it comes to sole Parameter Values.

**Reply by Leigten** [reply]
I admit that this dance between decimal, binary, and hexadecimal numbers has given me quite a headache (in fact, it has worsened the one I already had, given my condition).
Anyway, from what I can simply see by moving the controls on the Waldorf Blofeld and reading from the MIDI input monitor window, the correct combination of values is : 
0,20,01,21,02,22,03,23,04,24
...which is actually what you get from the variable and expression settings made by dnaldoog.

**Comment by dobo365** [comment]
Info about Global variables:

**Ctrlr global variables**
Global variables are variables declared at panel level and stored as a panel property.
  
Property: panelGlobalVariables
There are 4 series of 16 values:
─	k0..k9, ka..kf,
─	o1..o9,oa…of
─	p1..p9,pa…pf
─	n1..n9,na…nf

This is visible when you edit a sysex formula for example (right click on the cell of an individual value).
Syntax to set an individual global variable:
```
panel:setGlobalVariable(0, value) -> set k0
panel:setGlobalVariable (1, value) -> set k1
```
It is also possible to directly set a series of global variables at once by replacing the content of the panel property.

**Example 1**: m1 to m30 are representing some variables in your code/panel and you want to store them as global variables
`panel:setProperty("panelGlobalVariables",m1..":"..m2..":"..m3..":"..m4..":"..m5..":"..m6..":"..m7..":"..m8..":"..m9..":"..m10..":"..m11..":"..m12..":"..m13..":"..m14..":"..m15..":"..m16..":"..m17..":"..m18..":"..m19..":"..m20..":"..m21..":"..m22..":"..m23..":"..m24..":"..m25..":"..m26..":"..m27..":"..m28..":"..m29..":"..m30, false)`

**Example 2**: iUnitID, iMidiInChannel, iMidiOutChannel are 3 variables that you want to store as global variables
`panel:setProperty(“panelGlobalVariables”, iUnitID..”:”..iMidiInChannel..”:”..iMidiOutChannel..”:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0″,false)`

**Reply by dnaldoog** [reply]
I didn’t know about all this apart from `k0,k1 …` etc.  set and getGlobVariable() and setting it in the panel editor. 🙏

**Reply by Leigten** [reply]
What I would never get to is the right-click on the cell of an individual value by editing the sysex formula. 
As far as setting an individual global variable, whose syntax you describe, what I'm missing is where you actually make that setting.

**Reply by dnaldoog** [reply]
To set a global variable in the sysex where the value of the slider value  is `xx` just do `k0` etc, so something like `f0 3e 13 00 20 00 k1 3f xx f7`  &#x2015; so this would set byte 6 in the sysex message to global variable k1's value &#x2015; but in the case of the panel I sent you the global variable is already built into the expression.

**Reply by Leigten** [reply]
Thanks!
I had already found this information in other posts, I mean, that you can insert the variable directly into the sysex string.
What dobo365 says seems to be something else though:
panel:setGlobalVariable(0, value) -> set k0
It is a very specific syntax that I guess needs to be entered somewhere.....

**Reply by dnaldoog** [reply]
Yes that is lua right there - it’s lua’s setter for k0.

 if you have little to no experience with programming  then it will probably be quite a tough learning curve, but the language itself is actually quite small and relatively easy compared to a lot of other languages. If you have a good read of the PIL lua manual I linked to, it’s really very informative.

there are also [online lua sandbox compilers](https://www.tutorialspoint.com/execute_lua_online.php) that you can experiment with.

**Reply by Leigten** [reply]
my programming experiences go back to the Jurassic period, when PCs had a black screen and a flashing green prompt

**Comment by dnaldoog** [comment]
Hope you are feeling better @Leigten 

Here is a link about [expressions in Ctrlr](https://ctrlr.org/expressions-in-ctrlr-2/).

You can set a global here:

![global variable setting](https://github.com/RomanKubiak/ctrlr/assets/76831683/336bf94d-c519-48f3-b350-a31fe6b629b1)

To set the sysex that Tedjuh spotted I would just multiply the normal/single values by 32 and add to the other fields using the same system as before. 

At some point you'll probably want to learn [ _lua_](https://www.lua.org/pil/) though 😒.




So the expression for the normal/single `uiCombo` would be `modulatorValue*32+global.k3` and for the other values (trigger?) `modulatorValue+(global.k2*32)`.

See attached.

----

[Waldorf Blofeld_1_0_Test_dnaldoog3.zip](https://github.com/RomanKubiak/ctrlr/files/14771369/Waldorf.Blofeld_1_0_Test_dnaldoog3.zip)

**Reply by Leigten** [reply]
I am slightly better, thank you!
So, let me see if I understand....
From what dobo365 explained, there are these 4 sets of variables:
─ k0..k9, ka..kf,
─ o1..o9,oa..of
─ p1..p9,pa...pf
─ n1..n9,na...nf

In the attached image, we see the Modulator "Trigger" settings, which determine whether the envelope is of the "normal" or "single" type. In section A (Modulator attached to global variable) you have attached the modulatorValue to the variable k2, setting the value to 2.
I have seen that usually this value is set to -1, so I guess 0 points to k0, 1 to k1, 2 to k2 and so on....
The value taken by the variable k2 is the one in section B (Current modulator value), which in this case can have value 0 or 1.
As for Modulator "Mode" (i.e., Envelope Mode), the modulatorValue is hooked to the variable k3, which takes the value in "Current modulator value" and can vary from 0 to 4.
Your new version of the panel works perfectly, I obviously had to enter the correct sysex string in the SysEx Formula field.
So, thanks to the intervention of 3 people, Ted-juh, dnaldoog and dobo365, I got the solution to my problem. 
The key aspect is not so much the solution to the problem itself, but the information needed to get to the solution. 
All the ones I got here are priceless and I thank very much those who spent a little of their time to give me what I got.
I can slowly finish my panel, which for now is a simplified version. I hope to be able to make a version that is also graphically pleasing and share it here with those who may ever find it useful....

Thank you all!
![global variable](https://github.com/RomanKubiak/ctrlr/assets/164783721/e6bcd8de-1aa8-4a27-ac88-9b686990770b)

**Reply by dnaldoog** [reply]
I think you have explained it back perfectly and well done, which means you understand it. 

With regard to Goodweather’s other global variables, I think those have to be set in lua ? but I have never really used them or even the ones I showed you, because sometimes I have found them not to be so stable, and usually I use lua and set global lua variables, but lua may be something for you to look at in the future; so for now these the system probably works perfectly.

p.s I only stumbled on that global variable setting about a year ago, it’s not obvious with its -1 as a default value.

Getting into Ctrlr can be a very frustrating thing as there is no real documentation, but when you do discover an Easter egg so to speak, it’s quite thrilling! Or maybe I need to get a life?

**Reply by Leigten** [reply]
I agree, many things are not so obvious and the lack of documentation is really frustrating for Ctrlr beginners.
But when you get to know, to understand, things get exciting.
Do we need to get a life ? Our passions make our lives interesting.
Is your passion Ctrlr? Wonderful!
Is my passion music and synths? Wonderful!
Without passions, it would be a really dull life....

**Comment by dnaldoog** [comment]
In addition to Goodweather's guide on Ctrlr, please look at [this post ](https://github.com/RomanKubiak/ctrlr/discussions/509) - the panel there will give you a good idea of how to get into lua.

**Reply by Leigten** [reply]
Thank you very much John! 
I really don't know if I will ever get into LUA.
I only own two synths, and one of them has so many controls in its control panel that they absolutely beg to be used.
So my need was limited to the Blofeld. 
In any case, knowledge is always something valuable! 
In that post I read that it is only for Ctrlr version 5.3.201.
I currently have Windows version 5.4.52 and it is the third version I have installed. The previous two, which were higher versions, gave me some problems. Is there a recommended stable version of Ctrlr?
Thank you very much!

**Reply by dnaldoog** [reply]
Yes! only use version 5.3.201. It’s generally agreed that it is the most stable version.

Also there is CtrlrX which is being maintained by @damiensellier and is fast becoming very up to date and maybe going to supplant 5.3.201 if not already then soon.

**Reply by Leigten** [reply]
I had already heard about CtrlX.... Good to know, thanks!

**Comment by unityconstruct** [comment]
> At some point you'll probably want to learn [ _lua_](https://www.lua.org/pil/) though 😒.
@dnaldoog ... the exact reason I found Ctrlr 5 years ago, slapped together some mods, then ran into the lua wtf-zone and bailed for 5 years before sucking it up
